C7/Map/MapView.cs

```diff
@@ -0,0 +1,607 @@
+			MapUnit currentlySelected = game.CurrentlySelectedUnit;
+			foreach (MapUnit unit in units) {
+				if (unit == currentlySelected) {
+					selected = unit;
+				}
+				if (unit.HasPriorityAsDefender(bestDefender, currentlySelected)) {
+					bestDefender = unit;
+				}
+				if (game.animTracker.getUnitAppearance(unit).DeservesPlayerAttention()) {
+					interesting = unit;
+				}
+			}
+			// Prefer showing the selected unit, secondly show one doing a relevant animation, otherwise show the top defender
+			return selected ?? interesting ?? bestDefender;
+		}
+
+		public List<Tile> getVisibleTiles() {
+			List<Tile> tiles = new List<Tile>();
+			Rect2 bounds = game.camera.getVisibleWorld();
+			Vector2I topLeft = tilemap.LocalToMap(ToLocal(bounds.Position));
+			Vector2I bottomRight = tilemap.LocalToMap(ToLocal(bounds.End));
+			for (int x = topLeft.X - 1; x < bottomRight.X + 1; x++) {
+				for (int y = topLeft.Y - 1; y < bottomRight.Y + 1; y++) {
+					(int usX, int usY) = unstackedCoords(new Vector2I(x, y));
+					tiles.Add(data.map.tileAt(usX, usY));
+				}
+			}
+			return tiles;
+		}
+
+		public void updateAnimations() {
+			foreach (UnitSprite s in unitSprites.Values) {
+				s.Hide();
+			}
+			cursor.Hide();
+			foreach (Tile tile in getVisibleTiles()) {
+				MapUnit unit = selectUnitToDisplay(tile.unitsOnTile);
+				if (unit != MapUnit.NONE) {
+					animateUnit(tile, unit);
+				}
+			}
+		}
+
+		private void initializeTileMap() {
+			terrainTilemap = new TileMap();
+			terrainTilemapShadow = new TileMap();
+			terrainTileset = Civ3TerrainTileSet.Generate();
+			terrainTilemap.TileSet = terrainTileset;
+			terrainTilemap.Position += Vector2I.Right * (tileSize.X / 2);
+			terrainTilemapShadow.TileSet = terrainTileset;
+			terrainTilemapShadow.Position = terrainTilemap.Position + (Vector2I.Left * tileSize.X * width);
+
+			tilemap = new TileMap{ YSortEnabled = true };
+			tileset = TileSetLoader.LoadCiv3TileSet();
+			tilemap.TileSet = tileset;
+
+			// create tilemap layers
+			foreach (Layer layer in Enum.GetValues(typeof(Layer))) {
+				if (layer != Layer.Invalid) {
+					tilemap.AddLayer(layer.Index());
+					tilemap.SetLayerYSortEnabled(layer.Index(), true);
+					if (layer != Layer.FogOfWar) {
+ 						tilemap.SetLayerYSortEnabled(layer.Index(), true);
+ 					} else {
+ 						tilemap.SetLayerZIndex(layer.Index(), MapZIndex.FogOfWar);
+ 					}
+				}
+			}
+
+			tilemap.ZIndex = MapZIndex.Tiles;
+			AddChild(tilemap);
+			AddChild(terrainTilemap);
+			// AddChild(terrainTilemapShadow);
+		}
+
+		private void setTerrainTiles() {
+			string[] corners(int x, int y) {
+				string left = terrain[x, y];
+				string right = terrain[(x + 1) % width, y];
+				bool even = y % 2 == 0;
+				string top = "coast";
+				if (y > 0) {
+					top = even ? terrain[x, y - 1] : terrain[(x + 1) % width, y - 1];
+				}
+				string bottom = "coast";
+				if (y < height - 1) {
+					bottom = even ? terrain[x, y + 1] : terrain[(x + 1) % width, y + 1];
+				}
```

Comment:
Is it valid for these conditions to be false and the strings left as "coast"? What does that imply?

Reply:
TL;DR, explanation below: these conditions will always be false for tiles on non-wrapping edges. "coast" has no implication here since it's never visible, and is the cleanest default since it always gives _a_ visually correct terrain tile. 

civ terrain tiles represent the corners of 4 game tiles - so at the edges of the map, there is 1/4 of a terrain tile that is "off map". When choosing which terrain tile to use for the intersection of the 3 edge tiles, we need to choose a valid terrain tile combination (terrain tiles can only represent the intersection of 3 different terrain types, never 4). One of the following 2 predicates is always true of the terrain types of the corner of 3 tiles at an edge of the map:
A - one of those three tiles is coast => choosing coast for the 4th "fake, off map" tile is OK
B - two of those tiles have the same terrain type (and it is not coast), and the combination of terrain types will allow for the 4th "fake, off map" tile to be coast

If this were not a tilemap (ie. how the 3.x branch does it), we could instead sample from the corners of 4 different terrain tiles to draw the terrain underneath a game tile. Since we need to pick instead a terrain tile to draw under the joining corner of tiles, coast is a legal option which will give us a valid terrain tile to draw, and the fact that the 4th fake tile corner is hard coded to be coast is inconsequential since it's outside of the playable / visible map.